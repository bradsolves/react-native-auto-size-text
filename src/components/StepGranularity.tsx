import * as React from 'react';
import { NativeSyntheticEvent, Text, TextLayoutEventData } from 'react-native';

import { AutoSizeTextProps } from '../types';

const StepGranularity = (props: AutoSizeTextProps) => {
  const { fontSize, children, style, numberOfLines, granularity } = props;
  const [currentFont, setCurrentFont] = React.useState(fontSize);
  const handleResizing = (e: NativeSyntheticEvent<TextLayoutEventData>) => {
    const { lines } = e.nativeEvent;
    if (lines.length > (numberOfLines as number)) {
      setCurrentFont((currentFont as number) - (granularity as number));
    }
  };

  return (
    <Text
      numberOfLines={numberOfLines}
      style={[
        style,
        {
          fontSize: currentFont,
        },
      ]}
      onTextLayout={handleResizing}
    >
      {children}
    </Text>
  );
};

export default StepGranularity;
