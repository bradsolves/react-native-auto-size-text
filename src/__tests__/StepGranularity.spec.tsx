import { cleanup, render, waitFor } from '@testing-library/react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import StepGranularity from '../components/StepGranularity';

afterEach(cleanup);

describe('StepGranularity snapshot', () => {
  it('StepGranularity renders correctly', () => {
    const tree = renderer
      .create(
        <StepGranularity fontSize={14} granularity={10} numberOfLines={3}>
          Lorem impsum
        </StepGranularity>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('StepGranularity mode', () => {
  it('should pass the right props to Text component inside StepGranularity', async () => {
    const mockContent = 'Lorem ipsum';
    const mockFontSize = 64;
    const mockColor = 'red';

    const { getByTestId } = render(
      <StepGranularity
        style={{ color: mockColor }}
        fontSize={mockFontSize}
        numberOfLines={2}
        granularity={10}
      >
        {mockContent}
      </StepGranularity>
    );

    const textComponent = getByTestId('step-granularity');
    const content = textComponent.props.children;
    const numberOfLines = textComponent.props.numberOfLines;
    const adjustsFontSizeToFit = textComponent.props.adjustsFontSizeToFit;
    const fontSize = textComponent.props.style[1].fontSize;
    const color = textComponent.props.style[0].color;

    await waitFor(() => {
      expect(content).toBe(mockContent);
      expect(numberOfLines).toBeFalsy();
      expect(fontSize).toBe(mockFontSize);
      expect(color).toBe(mockColor);
      expect(adjustsFontSizeToFit).toBeFalsy();
    });
  });
});
