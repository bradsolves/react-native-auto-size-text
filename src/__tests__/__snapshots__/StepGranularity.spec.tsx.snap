// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepGranularity snapshot StepGranularity renders correctly 1`] = `
<Text
  onTextLayout={[Function]}
  style={
    Array [
      undefined,
      Object {
        "fontSize": 14,
      },
    ]
  }
  testID="step-granularity"
>
  Lorem impsum
</Text>
`;
